src/app/reporting/ReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,211 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/reporting/ReportSchedulerImpl.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+using Seconds16       = System::Clock::Seconds16;
+using Milliseconds32  = System::Clock::Milliseconds32;
+using Timeout         = System::Clock::Timeout;
+using Timestamp       = System::Clock::Timestamp;
+using ReadHandlerNode = ReportScheduler::ReadHandlerNode;
+
+/// @brief Callback called when the report timer expires to schedule an engine run regardless of the state of the ReadHandlers, as
+/// the engine already verifies that read handlers are reportable before sending a report
+static void ReportTimerCallback()
+{
+    InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleRun();
+}
+
+ReportSchedulerImpl::ReportSchedulerImpl(TimerDelegate * aTimerDelegate) : ReportScheduler(aTimerDelegate)
+{
+    VerifyOrDie(nullptr != mTimerDelegate);
+}
+
+/// @brief When a ReadHandler is added, register it, which will schedule an engine run
+void ReportSchedulerImpl::OnReadHandlerAdded(ReadHandler * aReadHandler)
+{
+    RegisterReadHandler(aReadHandler);
+}
+
+/// @brief When a ReadHandler becomes reportable, schedule, verifies if the min interval of a handleris elapsed. If not,
+/// reschedule the report to happen when the min interval is elapsed. If it is, schedule an engine run.
+void ReportSchedulerImpl::OnBecameReportable(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturn(nullptr != node);
+
+    Milliseconds32 newTimeout;
+    if (node->IsReportableNow())
+    {
+        // If the handler is reportable now, just schedule a report immediately
+        newTimeout = Milliseconds32(0);
+    }
+    else
+    {
+        // If the handler is not reportable now, schedule a report for the min interval
+        newTimeout = Milliseconds32(node->GetMinTimestamp().count() - mTimerDelegate->GetCurrentMonotonicTimestamp().count());
+    }
+
+    ScheduleReport(newTimeout, aReadHandler);
+}
+
+void ReportSchedulerImpl::OnReportSent(ReadHandler * apReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(apReadHandler);
+    VerifyOrReturn(nullptr != node);
+    // Schedule callback for max interval by computing the difference between the max timestamp and the current timestamp
+    node->SetIntervalsTimeStamp(apReadHandler);
+    Milliseconds32 newTimeout =
+        Milliseconds32(node->GetMaxTimestamp().count() - mTimerDelegate->GetCurrentMonotonicTimestamp().count());
```

Comment:
The fact that we use different now() calls for computing newTimeout and SetIntervalsTimeStamp is not great.  Followup please to not do that; it will lead to weird CI failures due to losing the timeslice between the calls on Linux/Mac, and might be somewhat expensive on other platforms.